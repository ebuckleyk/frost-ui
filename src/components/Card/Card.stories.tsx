import * as React from 'react';
import { Meta, StoryObj } from '@storybook/react';
import { BellRing, Check } from 'lucide-react';

import { cn } from '../../lib/utils';
import { Button } from '../Button';
import { Input } from '../Input';
import { Label } from '../Label';
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from '../Select';
import { Switch } from '../Switch';
import { Card, CardContent, CardDescription, CardFooter, CardHeader, CardTitle } from './Card';

const notifications = [
  {
    title: 'Your call has been confirmed.',
    description: '1 hour ago',
  },
  {
    title: 'You have a new message!',
    description: '1 hour ago',
  },
  {
    title: 'Your subscription is expiring soon!',
    description: '2 hours ago',
  },
];

type CardProps = React.ComponentProps<typeof Card>;
function CardDemo({ className, ...props }: CardProps) {
  return (
    <Card className={cn('w-[380px]', className)} {...props}>
      <CardHeader>
        <CardTitle>Notifications</CardTitle>
        <CardDescription>You have 3 unread messages.</CardDescription>
      </CardHeader>
      <CardContent className="grid gap-4">
        <div className=" flex items-center space-x-4 rounded-md border p-4">
          <BellRing />
          <div className="flex-1 space-y-1">
            <p className="text-sm font-medium leading-none">Push Notifications</p>
            <p className="text-sm text-muted-foreground">Send notifications to device.</p>
          </div>
          <Switch />
        </div>
        <div>
          {notifications.map((notification, index) => (
            <div key={index} className="mb-4 grid grid-cols-[25px_1fr] items-start pb-4 last:mb-0 last:pb-0">
              <span className="bg-sky-500 flex size-2 translate-y-1 rounded-full" />
              <div className="space-y-1">
                <p className="text-sm font-medium leading-none">{notification.title}</p>
                <p className="text-sm text-muted-foreground">{notification.description}</p>
              </div>
            </div>
          ))}
        </div>
      </CardContent>
      <CardFooter>
        <Button className="w-full">
          <Check className="mr-2 size-4" /> Mark all as read
        </Button>
      </CardFooter>
    </Card>
  );
}
function CardWithFormDemo({ className, ...props }: CardProps) {
  return (
    <Card className={cn('w-[380px]', className)} {...props}>
      <CardHeader>
        <CardTitle>Create project</CardTitle>
        <CardDescription>Deploy your new project in one-click.</CardDescription>
      </CardHeader>
      <CardContent>
        <form>
          <div className="grid w-full items-center gap-4">
            <div className="flex flex-col space-y-1.5">
              <Label htmlFor="name">Name</Label>
              <Input id="name" placeholder="Name of your project" />
            </div>
            <div className="flex flex-col space-y-1.5">
              <Label htmlFor="framework">Framework</Label>
              <Select>
                <SelectTrigger id="framework">
                  <SelectValue placeholder="Select" />
                </SelectTrigger>
                <SelectContent position="popper">
                  <SelectItem value="next">Next.js</SelectItem>
                  <SelectItem value="sveltekit">SvelteKit</SelectItem>
                  <SelectItem value="astro">Astro</SelectItem>
                  <SelectItem value="nuxt">Nuxt.js</SelectItem>
                </SelectContent>
              </Select>
            </div>
          </div>
        </form>
      </CardContent>
      <CardFooter className="flex justify-between">
        <Button variant="outline">Cancel</Button>
        <Button>Deploy</Button>
      </CardFooter>
    </Card>
  );
}

const meta: Meta<typeof Card> = {
  component: Card,
  subcomponents: { CardContent, CardDescription, CardFooter, CardHeader, CardTitle },
};

export default meta;

type Story = StoryObj<typeof Card>;

export const Demo: Story = {
  render: CardDemo,
};

export const CardWithForm: Story = {
  render: CardWithFormDemo,
};
