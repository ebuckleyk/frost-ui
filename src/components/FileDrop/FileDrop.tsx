import * as React from 'react';
import Dropzone, { DropzoneOptions, DropzoneRootProps, useDropzone } from 'react-dropzone';

import { cn } from '@/lib/utils';

// export interface FileDropProps extends DropzoneProps {}

// export interface ServerFile extends FileWithPath {
//   id?: string;
// }

const FILEDROP_NAME = 'FileDrop';

export type FileDropProps = React.ComponentPropsWithoutRef<typeof Dropzone> & DropzoneOptions & DropzoneRootProps;

const FileDrop = React.forwardRef<HTMLDivElement, FileDropProps>(({ className, children, ...props }, ref) => {
  const { getRootProps, isDragAccept, isFocused, isDragReject } = useDropzone(props);
  const style = React.useMemo(
    () =>
      `${isDragAccept ? 'border-green-400/30' : ''} ${isFocused ? 'border-input' : ''} ${
        isDragReject ? 'border-red-400/30' : ''
      }`,
    [isFocused, isDragAccept, isDragReject],
  );

  const dragOpacity = React.useMemo(() => (isDragAccept ? '' : '/30'), [isDragAccept]);

  return <div {...getRootProps(props)} ref={ref} children={children} />;
  return (
    <div>
      <div
        ref={ref}
        className={cn(
          `bg-primary-foreground${dragOpacity} flex-col items-center border-2 border-dashed border-primary-foreground${dragOpacity} rounded transition .24s ease-in-out flex p-5 ${style}`,
          className,
        )}
        {...getRootProps(props)}
        children={children}
      />
    </div>
  );
});

FileDrop.displayName = 'FileDrop';

const FileDropArea = React.forwardRef<HTMLDivElement, DropzoneOptions & React.HTMLAttributes<HTMLDivElement>>(
  ({ children, className, ...props }) => {
    const { getInputProps, isDragAccept, isFocused, isDragReject } = useDropzone(props);

    const style = React.useMemo(
      () =>
        `${isDragAccept ? 'border-green-400/30' : ''} ${isFocused ? 'border-input' : ''} ${
          isDragReject ? 'border-red-400/30' : ''
        }`,
      [isFocused, isDragAccept, isDragReject],
    );

    const dragOpacity = React.useMemo(() => (isDragAccept ? '' : '/30'), [isDragAccept]);
    console.log('rendered');
    return (
      <div
        className={cn(
          `bg-primary-foreground${dragOpacity} flex-col items-center border-2 border-dashed border-primary-foreground${dragOpacity} rounded transition .24s ease-in-out flex p-5 ${style}`,
          className,
        )}
      >
        <input {...getInputProps()} />
        {children}
      </div>
    );
  },
);

FileDropArea.displayName = 'FileDropArea';

const FileDropAreaText = React.forwardRef<HTMLParagraphElement, React.ParamHTMLAttributes<HTMLParagraphElement>>(
  ({ className, ...props }, ref) => {
    return <p ref={ref} className={cn(className)} {...props} />;
  },
);

FileDropAreaText.displayName = 'FileDropAreaText';

export { FileDrop, FileDropArea, FileDropAreaText };
