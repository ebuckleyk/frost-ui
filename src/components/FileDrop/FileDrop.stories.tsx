import * as React from 'react';
import { zodResolver } from '@hookform/resolvers/zod';
import { Meta, StoryObj } from '@storybook/react';
import { ErrorCode } from 'react-dropzone';
import { useForm } from 'react-hook-form';
import * as z from 'zod';

import { Button } from '../Button';
import { Card, CardContent, CardDescription, CardHeader, CardTitle } from '../Card';
import { Form, FormControl, FormField, FormItem, FormLabel, FormMessage } from '../Form';
import { ScrollArea } from '../ScrollArea';
import { toast } from '../Toast';
import { FileDrop, FileDropArea, FileDropAreaText, FileDropFileArea } from './FileDrop';

function FileDropDemo() {
  return (
    <Card className="p-5">
      <CardHeader>
        <CardTitle>Upload File</CardTitle>
        <CardDescription>Drag & Drop or Click Browse to Upload File</CardDescription>
      </CardHeader>
      <CardContent>
        <FileDrop>
          <FileDropArea>
            <FileDropAreaText className="cursor-pointer text-slate-700/30 underline">Browse</FileDropAreaText>
          </FileDropArea>
          <FileDropFileArea />
        </FileDrop>
      </CardContent>
    </Card>
  );
}

function AcceptFileTypeFileDropDemo() {
  return (
    <Card className="p-5">
      <CardHeader>
        <CardTitle>Upload File</CardTitle>
        <CardDescription>Drag & Drop or Click Browse to Upload File</CardDescription>
      </CardHeader>
      <CardContent>
        <FileDrop
          accept={{
            'text/html': ['.html', '.htm'],
          }}
        >
          <FileDropArea>
            <FileDropAreaText className="text-slate-700/30">Only *.html and *.htm files accepted.</FileDropAreaText>
          </FileDropArea>
        </FileDrop>
      </CardContent>
    </Card>
  );
}

const file1 = new File(['file'], 'testfile.pdf', {
  type: 'application/pdf',
});
const file2 = new File(['file'], '123.pdf', {
  type: 'application/pdf',
});

const generateFiles = (count: number): File[] => {
  const list: File[] = [];
  for (let i = 0; i < count; i++) {
    list.push(
      new File(['file'], `file ${i}.pdf`, {
        type: 'application/pdf',
      }),
    );
  }
  return list;
};

function InitialFilesFileDropDemo() {
  const [state] = React.useState<File[]>([file1, file2]);
  return (
    <Card className="p-5">
      <CardHeader>
        <CardTitle>Upload File</CardTitle>
        <CardDescription>Drag & Drop or Click Browse to Upload File</CardDescription>
      </CardHeader>
      <CardContent>
        <FileDrop files={state}>
          <FileDropArea>
            <FileDropAreaText className="text-slate-700/30">See preloaded files.</FileDropAreaText>
          </FileDropArea>
          <FileDropFileArea />
        </FileDrop>
      </CardContent>
    </Card>
  );
}

function DisabledFileDropDemo() {
  const [state] = React.useState<File[]>([file1, file2]);
  return (
    <Card className="p-5">
      <CardHeader>
        <CardTitle>Upload File</CardTitle>
        <CardDescription>Field and Files are readonly. Files are still selectable</CardDescription>
      </CardHeader>
      <CardContent>
        <FileDrop disabled files={state}>
          <FileDropArea>
            <FileDropAreaText className="text-slate-700/30">See preloaded files.</FileDropAreaText>
          </FileDropArea>
          <FileDropFileArea />
        </FileDrop>
      </CardContent>
    </Card>
  );
}

function FileDropWithScrollAreaDemo() {
  const [state] = React.useState<File[]>(generateFiles(10));
  return (
    <Card className="p-5">
      <CardHeader>
        <CardTitle>Upload File</CardTitle>
        <CardDescription>Field and Files are readonly. Files are still selectable</CardDescription>
      </CardHeader>
      <CardContent>
        <FileDrop disabled files={state}>
          <FileDropArea>
            <FileDropAreaText className="text-slate-700/30">See preloaded files.</FileDropAreaText>
          </FileDropArea>
          <ScrollArea className="h-36">
            <FileDropFileArea />
          </ScrollArea>
        </FileDrop>
      </CardContent>
    </Card>
  );
}

const formSchema = z.object({
  files: z.instanceof(File).array().min(1, 'Must add at least one file').max(2, 'No more than 2 files allowed.'),
});

function FileDropFormDemo() {
  const form = useForm<z.infer<typeof formSchema>>({
    resolver: zodResolver(formSchema),
    defaultValues: {
      files: [],
    },
  });

  function onSubmit(values: z.infer<typeof formSchema>) {
    toast({
      title: 'You submitted',
      description: `${values.files.length} file(s)`,
    });
  }
  return (
    <Form {...form}>
      <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-8">
        <FormField
          control={form.control}
          name="files"
          render={({ field }) => {
            return (
              <FormItem>
                <FormLabel>Files</FormLabel>
                <FormControl>
                  <FileDrop
                    {...field}
                    maxFiles={2}
                    onDropRejected={(files) => {
                      const messages: string[] = [];

                      files.forEach((file) => {
                        const msgs: string[] = [];
                        file.errors.forEach((m) => {
                          switch (m.code) {
                            case ErrorCode.FileInvalidType:
                            case ErrorCode.FileTooLarge:
                            case ErrorCode.FileTooSmall: {
                              msgs.push(m.message);
                              break;
                            }
                          }
                        });
                        const em = `${file.file.name}: ${msgs.join(' ')}`;
                        messages.push(em);
                      });
                      form.setError('files', { message: messages.join('\n') });
                    }}
                  >
                    <FileDropArea>
                      <FileDropAreaText className="text-slate-700/30">Add files to form.</FileDropAreaText>
                    </FileDropArea>
                    <FileDropFileArea />
                  </FileDrop>
                </FormControl>
                <FormMessage />
              </FormItem>
            );
          }}
        />
        <Button type="submit">Submit</Button>
      </form>
    </Form>
  );
}

type ComponentType = React.ComponentProps<typeof FileDrop>;
const meta: Meta<ComponentType> = {
  component: FileDrop,
  render: FileDropDemo,
};

export default meta;

type Story = StoryObj<ComponentType>;
export const Demo: Story = {};

export const AcceptFileTypesDemo: Story = {
  render: AcceptFileTypeFileDropDemo,
};

export const InitFileDrop: Story = {
  render: InitialFilesFileDropDemo,
};

export const DisabledFileDrop: Story = {
  render: DisabledFileDropDemo,
};

export const FileDropForm: Story = {
  render: FileDropFormDemo,
};

export const FileDropWithScrollArea: Story = {
  render: FileDropWithScrollAreaDemo,
};
