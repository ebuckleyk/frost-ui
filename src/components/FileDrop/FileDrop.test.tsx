import * as React from 'react';
import { render } from '@testing-library/react';

import { Card, CardContent, CardDescription, CardHeader, CardTitle } from '../Card';
import { FileDrop, FileDropArea, FileDropAreaText, FileDropFileArea } from './FileDrop';

function Component() {
  return (
    <Card className="p-5">
      <CardHeader>
        <CardTitle>Upload File</CardTitle>
        <CardDescription>Drag & Drop or Click Browse to Upload File</CardDescription>
      </CardHeader>
      <CardContent>
        <FileDrop>
          <FileDropArea>
            <FileDropAreaText className="cursor-pointer text-slate-700/30 underline">Browse</FileDropAreaText>
          </FileDropArea>
          <FileDropFileArea />
        </FileDrop>
      </CardContent>
    </Card>
  );
}

describe('FileDrop', () => {
  it('should render FileDrop component to match snapshot', () => {
    const result = render(<Component />);
    expect(result).toMatchSnapshot();
  });
});
