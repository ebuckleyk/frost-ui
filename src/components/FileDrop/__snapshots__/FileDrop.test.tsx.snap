// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileDrop > should render FileDrop component to match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="rounded-xl border bg-card/60 text-card-foreground shadow-md backdrop-blur-md backdrop-saturate-200 p-5"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
        >
          <h3
            class="font-semibold leading-none tracking-tight"
          >
            Upload File
          </h3>
          <p
            class="text-sm text-muted-foreground"
          >
            Drag & Drop or Click Browse to Upload File
          </p>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class=""
          >
            <div
              class="bg-primary-foreground/30 flex-col items-center border-2 border-dashed border-primary-foreground/30 rounded transition .24s ease-in-out flex p-5"
              role="presentation"
              tabindex="0"
            >
              <input
                multiple=""
                style="border: 0px; clip: rect(0px, 0px, 0px, 0px); clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
                tabindex="-1"
                type="file"
              />
              <p
                class="cursor-pointer text-slate-700/30 underline"
              >
                Browse
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="rounded-xl border bg-card/60 text-card-foreground shadow-md backdrop-blur-md backdrop-saturate-200 p-5"
    >
      <div
        class="flex flex-col space-y-1.5 p-6"
      >
        <h3
          class="font-semibold leading-none tracking-tight"
        >
          Upload File
        </h3>
        <p
          class="text-sm text-muted-foreground"
        >
          Drag & Drop or Click Browse to Upload File
        </p>
      </div>
      <div
        class="p-6 pt-0"
      >
        <div
          class=""
        >
          <div
            class="bg-primary-foreground/30 flex-col items-center border-2 border-dashed border-primary-foreground/30 rounded transition .24s ease-in-out flex p-5"
            role="presentation"
            tabindex="0"
          >
            <input
              multiple=""
              style="border: 0px; clip: rect(0px, 0px, 0px, 0px); clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
              tabindex="-1"
              type="file"
            />
            <p
              class="cursor-pointer text-slate-700/30 underline"
            >
              Browse
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
